include/zephyr/drivers/usb/udc_buf.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Buffers for USB device support
+ */
+
+#ifndef ZEPHYR_INCLUDE_UDC_BUF_H
+#define ZEPHYR_INCLUDE_UDC_BUF_H
+
+#include <zephyr/net/buf.h>
+
+#if defined(CONFIG_DCACHE) && !defined(CONFIG_NOCACHE_MEMORY)
+/*
+ * Here we try to get DMA-safe buffers, but we lack a consistent source of
+ * information about data cache properties, such as line cache size, and a
+ * consistent source of information about what part of memory is DMA'able.
+ * For now, we simply assume that all available memory is DMA'able and use
+ * Kconfig option DCACHE_LINE_SIZE for alignment and granularity.
+ */
+#define UDC_BUF_ALIGN		CONFIG_DCACHE_LINE_SIZE
+#define UDC_BUF_GRANULARITY	CONFIG_DCACHE_LINE_SIZE
+#else
+/* Defaults alignment and granularity if there is not data cache */
+#define UDC_BUF_ALIGN		sizeof(void *)
+#define UDC_BUF_GRANULARITY	sizeof(void *)
+#endif
+
+#define UDC_K_HEAP_DEFINE_NO_CACHE(name, size)						\
+	K_HEAP_DEFINE_NOCACHE(name, size)
+
+#define UDC_K_HEAP_DEFINE_DMAABLE(name, size)						\
+	K_HEAP_DEFINE(name, size)
+
+#define UDC_K_HEAP_DEFINE(name, size)							\
+	COND_CODE_1(CONFIG_NOCACHE_MEMORY,						\
+		    (UDC_K_HEAP_DEFINE_NO_CACHE(name, size)),				\
+		    (UDC_K_HEAP_DEFINE_DMAABLE(name, size)))
```

Comment:
The names are a bit confusing. Both NO_CACHE and DMAABLE has to be in memory region accessible by DMA. The difference is whether the memory has to be invalidated/flushed by D-Cache or not.

For the time being I would go without the intermediate `UDC_K_HEAP_DEFINE_NO_CACHE` and `UDC_K_HEAP_DEFINE_DMAABLE` and just use the `K_HEAP_DEFINE_NOCACHE` and `K_HEAP_DEFINE` directly in the `COND_CODE_1` here.

Reply:
NOCACHE on its own is clear, but because the other one is DMAABLE - it infers that the NOCACHE is not necessarily DMA-able.